ClassCastException when handling Avro Dates and Timestamps in ClickHouse Sink Connector

### Describe the bug
When using the Avro Converter with different date formats, particularly 'date' and 'timestamp-millis', the sink throws a `ClassCastException` error when used with Clickhouse types 'Date' and 'DateTime'. This happens because the sink attempts to cast a Date object to an Integer.
 'timestamp-millis' in combination with Clickhouse 'DateTime64' works just fine. 

### Steps to reproduce
1. Set up a Kafka Connect instance with the Avro Converter and ClickHouse Sink Connector.
2. Configure the Avro schema to include fields with 'date' and 'timestamp-millis' logical types.
3. Produce messages to Kafka with dates and timestamps according to the Avro schema.
4. Start the Kafka Connect instance to consume these messages and write them to ClickHouse.

### Expected behaviour
The Avro dates and timestamps should be properly converted and written into ClickHouse without causing a `ClassCastException`.


### Error log
```Java
java.lang.RuntimeException: java.lang.ClassCastException: class java.util.Date cannot be cast to class java.lang.Integer (java.util.Date and java.lang.Integer are in module java.base of loader 'bootstrap')
 	at com.clickhouse.kafka.connect.util.Utils.handleException(Utils.java:87)
 	at com.clickhouse.kafka.connect.sink.db.ClickHouseWriter.doInsert(ClickHouseWriter.java:143)
	at com.clickhouse.kafka.connect.sink.processing.Processing.doInsert(Processing.java:45)
	at com.clickhouse.kafka.connect.sink.processing.Processing.doLogic(Processing.java:102)
	at com.clickhouse.kafka.connect.sink.ProxySinkTask.put(ProxySinkTask.java:80)
 	at com.clickhouse.kafka.connect.sink.ClickHouseSinkTask.put(ClickHouseSinkTask.java:60)
 	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:587)
 	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:336)
 	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:237)
 	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:206)
	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:202)
 	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:257)
 	at org.apache.kafka.connect.runtime.isolation.Plugins.lambda$withClassLoader$1(Plugins.java:181)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
 Caused by: java.lang.ClassCastException: class java.util.Date cannot be cast to class java.lang.Integer (java.util.Date and java.lang.Integer are in module java.base of loader 'bootstrap')
```

### Configuration
#### Environment
* Confluent stack 7.4.1
* Kafka Connect configuration: Avro Converter, ClickHouse Sink Connector

#### ClickHouse server
* ClickHouse Server version: 23.5
* `CREATE TABLE` statements for tables involved: 
```SQL
CREATE TABLE test
(
    test_date Nullable(DATE),
    test_datetime Nullable(DATETIME)
) ENGINE = MergeTree()
      ORDER BY test_date
```

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
@@ -15,6 +15,7 @@ import com.clickhouse.kafka.connect.util.Mask;
 import com.clickhouse.kafka.connect.util.Utils;
 import com.google.gson.Gson;
 import com.google.gson.reflect.TypeToken;
+import java.util.concurrent.TimeUnit;
 import org.apache.kafka.connect.data.Field;
 import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.connect.data.Struct;

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
@@ -190,7 +190,13 @@ public class ClickHouseWriter implements DBWriter{
         switch (type) {
             case Date:
                 if (value.getFieldType().equals(Schema.Type.INT32)) {
-                    BinaryStreamUtils.writeUnsignedInt16(stream, (Integer) value.getObject());
+                    if (value.getObject().getClass().getName().endsWith(".Date")) {
+                        Date date = (Date)value.getObject();
+                        int timeInDays = (int) TimeUnit.MILLISECONDS.toDays(date.getTime());
+                        BinaryStreamUtils.writeUnsignedInt16(stream, timeInDays);
+                    } else {
+                        BinaryStreamUtils.writeUnsignedInt16(stream, (Integer) value.getObject());
+                    }
                 } else {
                     unsupported = true;
                 }
@@ -199,8 +205,8 @@ public class ClickHouseWriter implements DBWriter{
                 if (value.getFieldType().equals(Schema.Type.INT32)) {
                     if (value.getObject().getClass().getName().endsWith(".Date")) {
                         Date date = (Date)value.getObject();
-                        int time = (int)date.getTime();
-                        BinaryStreamUtils.writeInt32(stream, time);
+                        int timeInDays = (int) TimeUnit.MILLISECONDS.toDays(date.getTime());
+                        BinaryStreamUtils.writeInt32(stream, timeInDays);
                     } else {
                         BinaryStreamUtils.writeInt32(stream, (Integer) value.getObject());
                     }
@@ -210,7 +216,13 @@ public class ClickHouseWriter implements DBWriter{
                 break;
             case DateTime:
                 if (value.getFieldType().equals(Schema.Type.INT64)) {
-                    BinaryStreamUtils.writeUnsignedInt32(stream, (Long) value.getObject());
+                    if (value.getObject().getClass().getName().endsWith(".Date")) {
+                        Date date = (Date)value.getObject();
+                        long time = date.getTime();
+                        BinaryStreamUtils.writeUnsignedInt32(stream, time);
+                    } else {
+                        BinaryStreamUtils.writeUnsignedInt32(stream, (Long) value.getObject());
+                    }
                 } else {
 
                     unsupported = true;

--- a/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
+++ b/src/main/java/com/clickhouse/kafka/connect/sink/db/ClickHouseWriter.java
@@ -216,16 +216,16 @@ public class ClickHouseWriter implements DBWriter{
                 }
                 break;
             case DateTime:
-                if (value.getFieldType().equals(Schema.Type.INT64)) {
+                if (value.getFieldType().equals(Schema.Type.INT32) || value.getFieldType().equals(Schema.Type.INT64)) {
                     if (value.getObject().getClass().getName().endsWith(".Date")) {
                         Date date = (Date)value.getObject();
-                        long time = date.getTime();
-                        BinaryStreamUtils.writeUnsignedInt32(stream, time);
+                        long epochSecond = date.toInstant().getEpochSecond();
+                        BinaryStreamUtils.writeUnsignedInt32(stream, epochSecond);
                     } else {
                         BinaryStreamUtils.writeUnsignedInt32(stream, (Long) value.getObject());
                     }
                 } else {
-
+                    
                     unsupported = true;
                 }
                 break;

--- a/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
+++ b/src/test/java/com/clickhouse/kafka/connect/sink/ClickHouseSinkTaskWithSchemaTest.java
@@ -396,6 +396,8 @@ public class ClickHouseSinkTaskWithSchemaTest {
                 .field("timestamp_date", Timestamp.SCHEMA)
                 .field("time_int32" , Time.SCHEMA)
                 .field("time_date32" , Time.SCHEMA)
+                .field("date_date", Time.SCHEMA)
+                .field("datetime_date", Timestamp.SCHEMA)
                 .build();
 
 
@@ -421,6 +423,8 @@ public class ClickHouseSinkTaskWithSchemaTest {
                     .put("timestamp_date",  new Date(System.currentTimeMillis()))
                     .put("time_int32", new Date(System.currentTimeMillis()))
                     .put("time_date32", new Date(System.currentTimeMillis()))
+                    .put("date_date", new Date(System.currentTimeMillis()))
+                    .put("datetime_date", new Date(System.currentTimeMillis()))
                     ;
 
 
@@ -634,7 +638,7 @@ public class ClickHouseSinkTaskWithSchemaTest {
 
         String topic = "support-dates-table-test";
         dropTable(chc, topic);
-        createTable(chc, topic, "CREATE TABLE `%s` ( `off16` Int16, date_number Nullable(Date), date32_number Nullable(Date32), datetime_number DateTime, datetime64_number DateTime64, timestamp_int64 Int64, timestamp_date DateTime64, time_int32 Int32, time_date32 Date32 ) Engine = MergeTree ORDER BY off16");
+        createTable(chc, topic, "CREATE TABLE `%s` ( `off16` Int16, date_number Nullable(Date), date32_number Nullable(Date32), datetime_number DateTime, datetime64_number DateTime64, timestamp_int64 Int64, timestamp_date DateTime64, time_int32 Int32, time_date32 Date32, date_date Date, datetime_date DateTime ) Engine = MergeTree ORDER BY off16");
         // https://github.com/apache/kafka/blob/trunk/connect/api/src/test/java/org/apache/kafka/connect/data/StructTest.java#L95-L98
         Collection<SinkRecord> sr = createDateType(topic, 1);
 

